src/peft/tuners/lora/eva.py

```
@@ -0,0 +1,505 @@
+                handle.remove()
+                handle = None
+                convergence_dict[name] = True
+                continue
+            # if a layer has switched from converged to not converged in the current step
+            elif convergence_dict[name] and not converged:
+                module = model.get_submodule(name)
+                handle = module.register_forward_hook(hook)
+                convergence_dict[name] = False
+            hook.indices = indices.T.unbind()
+            hooks[name] = (hook, handle)
+
+        if show_progress_bar:
+            layer_converged = list(convergence_dict.values()) + [
+                convergence_dict[v] for v in equal_inputs_map.values()
+            ]
+            pbar.set_description(f"{sum(layer_converged)}/{len(layer_converged)} layers have converged")
+
+        if all(convergence_dict.values()):
+            break
+
+        forward_fn(model, move_inputs_to_device(inputs, device))
+
+        # in case some hooks have to skip the svd calculation because the number of tokens is less than the number of components
+        if not all(hasattr(h[0].svd, "components_") for h in hooks.values()):
+            continue
+
+        rank_dist = _get_rank_distribution(hooks, layer_hook_map, equal_inputs_map, rank_budget, max_components)
+
+    # check all svd hooks have been removed
+    remaining_svd_hooks = {
+        n for n, m in model.named_modules() for v in m._forward_hooks.values() if isinstance(v, SVDHook)
+    }
+    if len(remaining_svd_hooks) > 0:
+        raise ValueError(
+            f"All SVD hooks should have been removed but the following layers still have active hooks: {remaining_svd_hooks}"
+        )
+
+    eva_state_dict = {}
+    for name, rank in rank_dist.items():
+        if rank == 0:
+            continue
+        hook = hooks[layer_hook_map[name]][0]
+        assert torch.all(hook.converged[:rank])  # this should never happen because we check for convergence
```

Let's raise a proper error here.